src/circuit/num.rs

```diff
@@ -332,14 +325,19 @@ impl<E: Engine> AllocatedNum<E> {
             }
         )?;
 
+        // (1-c)(a) + (c)(b) = r
+        // (1-c)(b) + (c)(a) = s
+        // a - ca + cb - r = b - cb + ca - s
+        // c(2b - 2a) = (r - s) + (b - a)
```

Comment:
I don't think this can work. `r` and `s` are not linearly independent. Example: if `c = 0`, the equation reduces to `s + a = r + b` which does have solution `a = r`/`b = s` but also has solution `a = r+1`/`b = s+1`.

The only way to optimize this into one constraint is if the function returns a linear combination of elements, but `AllocatedNum` doesn't support this. `Num` does though. However, the code that will unpack this for use in the Merkle tree is implemented over `AllocatedNum` so it would require a bit of refactoring to get this right. Not sure it's worth it for the few constraints it ends up saving. (Probably 32 constraints out of many tens of thousands).

Reply:
Ah, good catch.  In that case I'll close this for now.